sound/soc/sof/core.c

```diff
@@ -516,6 +530,13 @@ int snd_sof_device_remove(struct device *dev)
 	snd_sof_free_debug(sdev);
 	snd_sof_free_trace(sdev);
 
+	/*
+	 * Block until all children are powered. Otherwise
+	 * we may have races between child device resume flows
+	 * and platform device removal.
+	 */
+	device_for_each_child(sdev->dev, NULL, __force_rpm_active);
```

Comment:
I am not fully clear on the race condition.

The commit message seems to mean that the topology unload will result in the DSP being turned off.
But we have another step below with an snd_sof_remove() which as I understand it turns the DSP off?

In other words, is the race between resuming children/removing machine device, or resuming children/calling snd_sof_remove()?

Reply:
@plbossart Hmm, ok, I see what you mean. The race is between resuming children/removing machine device. But yeah, the commit message is confusing now. I used "DSP turned off" in commit message, as the topology unload will effectively do this (unloading the scheduler will actuall shut down core 0). So when snd_sof_remove() is finally called, DSP is no longer powered.

Makes sense?

Reply:
@kv2019i I am not sure that on topology unload the DSP is actually completely powered-off. Maybe core0 is powered off but the rest of the controller isn't
@ranj063 would need to comment, if they get power in the East Bay?

Reply:
@plbossart Whole ADSP is not powered off, but the cores are. But, but, it's true this shouldn't block codec drivers (or any other machine driver components). I debugged more and it seems this goes to:

[   64.352571]  soc_remove_link_components+0x6c/0xb0 [snd_soc_core]
[   64.352573]  snd_soc_unbind_card+0x9b/0xe0 [snd_soc_core]
[   64.352575]  snd_soc_unregister_card+0x1d/0x60 [snd_soc_core]
[   64.352577]  release_nodes+0x121/0x220
[   64.352579]  device_release_driver_internal+0xf0/0x1c0
[   64.352580]  bus_remove_device+0xd6/0x140
[   64.352580]  device_del+0x15f/0x380
[   64.352582]  platform_device_del.part.14+0xe/0x60
[   64.352583]  platform_device_unregister+0x17/0x30

... and there we go to codec driver. So maybe we need to fix this in the codec driver after all. I'll continue debug and report back.

Reply:
@plbossart OK, I uploaded a new version. I'll continue to polish tomorrow but wanted to get CI results for the new version already today. It survives my CFL test bench and is a much simpler fix.

@lyakh I think you'll like this version better :)

Reply:
This looks good to go now. Thanks for the review comments @plbossart  and @lyakh , good that the original version did not go in.